url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/ReflectionBeansTemplateBuilder.java

```diff
@@ -0,0 +1,210 @@
+import java.util.ArrayList;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.msgpack.annotation.Ignore;
+import org.msgpack.annotation.Index;
+import org.msgpack.annotation.NotNullable;
+import org.msgpack.annotation.Optional;
+import org.msgpack.packer.Packer;
+import org.msgpack.template.FieldOption;
+import org.msgpack.template.Template;
+import org.msgpack.template.TemplateRegistry;
+import org.msgpack.template.builder.beans.BeanInfo;
+import org.msgpack.template.builder.beans.IntrospectionException;
+import org.msgpack.template.builder.beans.Introspector;
+import org.msgpack.template.builder.beans.PropertyDescriptor;
+import org.msgpack.unpacker.Unpacker;
+
+/**
+ * Class for building java reflection template builder for java beans class.
+ * 
+ * @author takeshita
+ * 
+ */
+@SuppressWarnings({ "rawtypes" })
+public class ReflectionBeansTemplateBuilder extends ReflectionTemplateBuilder {
+
+    private static Logger LOG = Logger.getLogger(ReflectionBeansTemplateBuilder.class.getName());
+
+    static class ReflectionBeansFieldTemplate extends ReflectionFieldTemplate {
+        ReflectionBeansFieldTemplate(final FieldEntry entry) {
+            super(entry);
+        }
+
+        
+        public void write(Packer packer, Object v, boolean required) throws IOException {
+            packer.write(v);
+        }
+
+        
+        public Object read(Unpacker unpacker, Object to, boolean required) throws IOException {
+            Object o = unpacker.read(entry.getType());
+            entry.set(to, o);
+            return o;
+        }
+    }
+
+    public ReflectionBeansTemplateBuilder(TemplateRegistry registry) {
+        super(registry, null);
+    }
+
+    
+    public boolean matchType(Type targetType, boolean hasAnnotation) {
+        Class<?> targetClass = (Class<?>) targetType;
+        boolean matched = matchAtBeansClassTemplateBuilder(targetClass, hasAnnotation);
+        if (matched && LOG.isLoggable(Level.FINE)) {
+            LOG.fine("matched type: " + targetClass.getName());
+        }
+        return matched;
+    }
+
+    
+    protected ReflectionFieldTemplate[] toTemplates(FieldEntry[] entries) {
+        ReflectionFieldTemplate[] tmpls = new ReflectionFieldTemplate[entries.length];
+        for (int i = 0; i < entries.length; i++) {
+            FieldEntry e = entries[i];
+            Class<?> type = e.getType();
+            if (type.isPrimitive()) {
+                tmpls[i] = new ReflectionBeansFieldTemplate(e);
+            } else {
+                Template tmpl = registry.lookup(e.getGenericType());
+                tmpls[i] = new FieldTemplateImpl(e, tmpl);
+            }
+        }
+        return tmpls;
+    }
+
+    
+    public FieldEntry[] toFieldEntries(Class<?> targetClass, FieldOption implicitOption) {
+        BeanInfo desc;
+        try {
+            desc = Introspector.getBeanInfo(targetClass);
+        } catch (IntrospectionException e1) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)